### PR TITLE
dev-java/pdfbox: additional TEST_EXCLUDES for jdk-17

### PR DESCRIPTION
**Pull request by vaukai**

### CODE REVIEW

**Comment by vaukai** [comment]
All tests pass for:
- dev-java/fontbox: bump to 2.0.24
- dev-java/jbig2-imageio: new package
- dev-java/jai-imageio-core: new package
- dev-java/jai-imageio-jpeg2000: new package

Has no tests
- dev-java/java-diff-utils: new package

Still to resolve: (Tests run: 600,  Failures: 17)
- dev-java/pdfbox: bump to 2.0.24 - CVE-2021-31812

**Comment by fordfrog** [comment]
@vaukai thanks for the summary, that was really helpful. merged everything except `pdfbox`. thanks!

**Comment by vaukai** [comment]
Running tests as a regular user now works with  `PDFontTest`, `TestCOSBase` and `TestCOSNumber` excluded:
```
OK (578 tests)
```
```
JAVA_TEST_EXCLUDES=(
    # excluded according to
    # https://github.com/apache/pdfbox/blob/8876e8e1a0adbf619cef4638cc3cea073e3ca484/pdfbox/pom.xml#L123
    "org.apache.pdfbox.rendering.TestPDFToImage"

    # 1) testPDFBox3826(org.apache.pdfbox.pdmodel.font.PDFontTest)
    # java.lang.IllegalArgumentException: URI is not hierarchical
    #         at java.base/java.io.File.<init>(File.java:420)
    #         at org.apache.pdfbox.pdmodel.font.PDFontTest.testPDFBox3826(PDFontTest.java:141)
    "org.apache.pdfbox.pdmodel.font.PDFontTest" # Tests run: 9,  Failures: 1, java.lang.IllegalArgumentException: URI is not hierarchical

    # junit.framework.AssertionFailedError: Cannot instantiate test case: testIsSetDirect (java.lang.InstantiationException
    # junit.framework.AssertionFailedError: Cannot instantiate test case: testAccept (java.lang.InstantiationException
    # junit.framework.AssertionFailedError: Cannot instantiate test case: testGetCOSObject (java.lang.InstantiationException
    "org.apache.pdfbox.cos.TestCOSBase"

    # junit.framework.AssertionFailedError: Cannot instantiate test case: testFloatValue (java.lang.InstantiationException
    # junit.framework.AssertionFailedError: Cannot instantiate test case: testGet (java.lang.InstantiationException
    # junit.framework.AssertionFailedError: Cannot instantiate test case: testInvalidNumber (java.lang.InstantiationException
    # junit.framework.AssertionFailedError: Cannot instantiate test case: testLongValue (java.lang.InstantiationException
    # junit.framework.AssertionFailedError: Cannot instantiate test case: testIntValue (java.lang.InstantiationException
    # junit.framework.AssertionFailedError: Cannot instantiate test case: testLargeNumber (java.lang.InstantiationException
    # junit.framework.AssertionFailedError: Cannot instantiate test case: testDoubleValue (java.lang.InstantiationException
    # junit.framework.AssertionFailedError: Cannot instantiate test case: testIsSetDirect (java.lang.InstantiationException
    # junit.framework.AssertionFailedError: Cannot instantiate test case: testAccept (java.lang.InstantiationException
    # junit.framework.AssertionFailedError: Cannot instantiate test case: testGetCOSObject (java.lang.InstantiationException
    "org.apache.pdfbox.cos.TestCOSNumber"   # Tests run: 10,  Failures: 10
)
```
However, when testing as root there are more test failures:
```
!!! When you file a bug report, please include the following information:
        at java.base/sun.net.www.protocol.https.HttpsClient.New(HttpsClient.java:372)
        at java.base/sun.net.www.protocol.https.AbstractDelegateHttpsURLConnection.getNewHttpClient(AbstractDelegateHttpsURLConnection.java:203)
        at java.base/sun.net.www.protocol.http.HttpURLConnection.plainConnect0(HttpURLConnection.java:1187)
        at java.base/sun.net.www.protocol.http.HttpURLConnection.plainConnect(HttpURLConnection.java:1081)
        at java.base/sun.net.www.protocol.https.AbstractDelegateHttpsURLConnection.connect(AbstractDelegateHttpsURLConnection.java:189)
        at java.base/sun.net.www.protocol.http.HttpURLConnection.getInputStream0(HttpURLConnection.java:1592)
        at java.base/sun.net.www.protocol.http.HttpURLConnection.getInputStream(HttpURLConnection.java:1520)
        at java.base/sun.net.www.protocol.https.HttpsURLConnectionImpl.getInputStream(HttpsURLConnectionImpl.java:250)
        at java.base/java.net.URL.openStream(URL.java:1165)
        at org.apache.pdfbox.pdmodel.interactive.form.PDAcroFormGenerateAppearancesTest.test4086CharNotEncodable(PDAcroFormGenerateAppearancesTest.java:78)

FAILURES!!!
Tests run: 578,  Failures: 31
```

**Comment by vaukai** [comment]
```
JAVA_TEST_EXCLUDES=(
    # excluded according to upstream
    # https://github.com/apache/pdfbox/blob/8876e8e1a0adbf619cef4638cc3cea073e3ca484/pdfbox/pom.xml#L123
    "org.apache.pdfbox.rendering.TestPDFToImage"

    # junit.framework.AssertionFailedError: Cannot instantiate test case: testIsSetDirect (java.lang.InstantiationException
    # junit.framework.AssertionFailedError: Cannot instantiate test case: testAccept (java.lang.InstantiationException
    # junit.framework.AssertionFailedError: Cannot instantiate test case: testGetCOSObject (java.lang.InstantiationException
    "org.apache.pdfbox.cos.TestCOSBase"

    # junit.framework.AssertionFailedError: Cannot instantiate test case: testFloatValue (java.lang.InstantiationException
    # junit.framework.AssertionFailedError: Cannot instantiate test case: testGet (java.lang.InstantiationException
    # junit.framework.AssertionFailedError: Cannot instantiate test case: testInvalidNumber (java.lang.InstantiationException
    # junit.framework.AssertionFailedError: Cannot instantiate test case: testLongValue (java.lang.InstantiationException
    # junit.framework.AssertionFailedError: Cannot instantiate test case: testIntValue (java.lang.InstantiationException
    # junit.framework.AssertionFailedError: Cannot instantiate test case: testLargeNumber (java.lang.InstantiationException
    # junit.framework.AssertionFailedError: Cannot instantiate test case: testDoubleValue (java.lang.InstantiationException
    # junit.framework.AssertionFailedError: Cannot instantiate test case: testIsSetDirect (java.lang.InstantiationException
    # junit.framework.AssertionFailedError: Cannot instantiate test case: testAccept (java.lang.InstantiationException
    # junit.framework.AssertionFailedError: Cannot instantiate test case: testGetCOSObject (java.lang.InstantiationException
    "org.apache.pdfbox.cos.TestCOSNumber"   # Tests run: 10,  Failures: 10

    # following tests need network access, can pass when run by regular user
    "org.apache.pdfbox.pdmodel.font.PDFontTest" # Tests run: 9,  Failures: 2
    "org.apache.pdfbox.pdmodel.interactive.form.PDAcroFormFlattenTest"  # Tests run: 11,  Failures: 11
    "org.apache.pdfbox.pdmodel.interactive.form.PDAcroFormFromAnnotsTest"   # Tests run: 7,  Failures: 7
    "org.apache.pdfbox.pdmodel.interactive.form.PDAcroFormGenerateAppearancesTest"  # Tests run: 3,  Failures: 3
    "org.apache.pdfbox.pdmodel.interactive.form.PDAcroFormTest" # Tests run: 11,  Failures: 4
    "org.apache.pdfbox.pdmodel.interactive.form.PDFieldTreeTest"    # Tests run: 1,  Failures: 1
    "org.apache.pdfbox.pdmodel.interactive.form.TestRadioButtons"   # Tests run: 9,  Failures: 8
)
```

**Comment by vaukai** [comment]
`pdfbox-debugger` added which is needed for [pdfbox-tools](https://github.com/apache/pdfbox/blob/8876e8e1a0adbf619cef4638cc3cea073e3ca484/tools/src/main/java/org/apache/pdfbox/tools/PDFBox.java)

**Comment by vaukai** [comment]
All tests of pdfbox-tools pass.

**Comment by fordfrog** [comment]
@vaukai `dev-java/picocli` merged, thanks! wrt `pdfbox`, i'd like to review the failing tests before merge.

**Comment by vaukai** [comment]
Some network tests enabled (would 31 of 584 tests justify this setting?)
```
PROPERTIES="test_network"
RESTRICT="test"
```

**Comment by vaukai** [comment]
> @vaukai `dev-java/picocli` merged, thanks! wrt `pdfbox`, i'd like to review the failing tests before merge.

After setting `PROPERTIES="test_network"` the only test failures left are
```
public void testPDFBox3826() throws IOException, URISyntaxException
```
from `"org.apache.pdfbox.pdmodel.font.PDFontTest"`
and those strange 10 failures from
```
"org.apache.pdfbox.cos.TestCOSBase"
"org.apache.pdfbox.cos.TestCOSNumber"
```

File: dev-java/pdfbox/pdfbox-2.0.24.ebuild [hunk 1]
```diff
@@ -0,0 +1,266 @@
...
+
+	mkdir target/imgs || die
+	cp \
+		"${DISTDIR}"/PDFBOX-5196-lotus.jpg \
+		"${DISTDIR}"/PDFBOX-4184-032163.jpg \
+		"${DISTDIR}"/PDFBOX-4184-16bit.png \
+		"target/imgs" || die
+
+	java-pkg-simple_src_test
+	rm -r target/test-classes || die # avaoid to run previous tests again
```
Comment (fordfrog) [inline dev-java/pdfbox/pdfbox-2.0.24.ebuild:238]: a typo in the comment

File: dev-java/pdfbox/files/2.0.24-tests-ignore.patch [hunk 2]
```diff
@@ -0,0 +1,18 @@
...
+ import org.junit.Test;
++import org.junit.Ignore;
+ 
+ /**
+  * 
+@@ -133,6 +134,7 @@ public class PDFontTest
+      * @throws IOException
+      * @throws URISyntaxException
+      */
++    @Ignore
```
Comment (fordfrog) [inline dev-java/pdfbox/files/2.0.24-tests-ignore.patch:15]: please add a comment why it needs to be ignored

**Comment by vaukai** [comment]
Excluding one more test only for jdk-17 where it seems to want a newer mockito. 
```
Time: 54.081
There was 1 failure:
1) testIsEmbeddingPermittedMultipleVersions(org.apache.pdfbox.pdmodel.font.TestFontEmbedding)
java.lang.ExceptionInInitializerError
        at org.mockito.cglib.core.KeyFactory$Generator.generateClass(KeyFactory.java:167)
        at org.mockito.cglib.core.DefaultGeneratorStrategy.generate(DefaultGeneratorStrategy.java:25)
        at org.mockito.cglib.core.AbstractClassGenerator.create(AbstractClassGenerator.java:217)
        at org.mockito.cglib.core.KeyFactory$Generator.create(KeyFactory.java:145)
        at org.mockito.cglib.core.KeyFactory.create(KeyFactory.java:117)
        at org.mockito.cglib.core.KeyFactory.create(KeyFactory.java:109)
        at org.mockito.cglib.core.KeyFactory.create(KeyFactory.java:105)
        at org.mockito.cglib.proxy.Enhancer.<clinit>(Enhancer.java:71)
        at org.mockito.internal.creation.jmock.ClassImposterizer.createProxyClass(ClassImposterizer.java:85)
        at org.mockito.internal.creation.jmock.ClassImposterizer.imposterise(ClassImposterizer.java:62)
        at org.mockito.internal.creation.jmock.ClassImposterizer.imposterise(ClassImposterizer.java:56)
        at org.mockito.internal.creation.CglibMockMaker.createMock(CglibMockMaker.java:23)
        at org.mockito.internal.util.MockUtil.createMock(MockUtil.java:26)
        at org.mockito.internal.MockitoCore.mock(MockitoCore.java:51)
        at org.mockito.Mockito.mock(Mockito.java:1243)
        at org.mockito.Mockito.mock(Mockito.java:1120)
        at org.apache.pdfbox.pdmodel.font.TestFontEmbedding.testIsEmbeddingPermittedMultipleVersions(TestFontEmbedding.java:350)
        ... 25 trimmed
Caused by: java.lang.reflect.InaccessibleObjectException: Unable to make protected final java.lang.Class java.lang.ClassLoader.defineClass(java.lang.String,byte[],int,int,java.security.ProtectionDomain) throws java.lang.ClassFormatError accessible: module java.base does not "opens java.lang" to unnamed module @4dc690>
        at java.base/java.lang.reflect.AccessibleObject.checkCanSetAccessible(AccessibleObject.java:354)
        at java.base/java.lang.reflect.AccessibleObject.checkCanSetAccessible(AccessibleObject.java:297)
        at java.base/java.lang.reflect.Method.checkCanSetAccessible(Method.java:199)
        at java.base/java.lang.reflect.Method.setAccessible(Method.java:193)
        at org.mockito.cglib.core.ReflectUtils$2.run(ReflectUtils.java:57)
        at java.base/java.security.AccessController.doPrivileged(AccessController.java:318)
        at org.mockito.cglib.core.ReflectUtils.<clinit>(ReflectUtils.java:47)
        ... 43 more

FAILURES!!!
Tests run: 586,  Failures: 1
```

**Comment by fordfrog** [comment]
what is the status here?

**Comment by vaukai** [comment]
> what is the status here?

From my point of view it's ready to merge except you find additional test failures that didn't occur here.

The JAVA_TEST_EXCLUDES are either excluded upstream 
or just warnings like »1) warning(junit.framework.TestSuite$1)«
The test marked @Ignore via patch I'd like to keep for later (no other idea how to deal with it).

File: dev-java/pdfbox/pdfbox-2.0.24.ebuild [hunk 3]
```diff
@@ -0,0 +1,291 @@
...
+		https://issues.apache.org/jira/secure/attachment/13025718/lotus.jpg -> PDFBOX-5196-lotus.jpg
+	)"
+
+LICENSE="Apache-2.0"
+SLOT="0"
+KEYWORDS="~amd64 ~arm64 ~ppc64 ~x86"
+IUSE="+tools"
+
+PROPERTIES="test_network"
+RESTRICT="test"
```
Comment (fordfrog) [inline dev-java/pdfbox/pdfbox-2.0.24.ebuild:66]: shouldn't this be dropped? i got the impression that the tests are working as you either fixed the tests or excluded them. or did i get it wrong?

**Comment by fordfrog** [comment]
@vaukai thanks! :-)

**Pull request closed without merging**